accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between and (2) are disjoint (such as iOS and Windows), the
+  information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
+2. '{API}' has been deprecated since {OS} {OS version}.
+3. '{API}' has been removed since {OS} {OS version}.
+
+### Automatic suppression via platform guards
+
+Diagnostic (1) can be suppressed via a call to
+`RuntimeInformation.IsOSPlatformOrLater()`. This should work for for simple cases
+where the call is contained inside an `if` check but also when the check causes
+control flow to never reach the call:
+
+```C#
+public void Case1()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        AppleApi();
+}
+
+public void Case2()
+{
+    if (!RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        return;
+
+    AppleApi();
+}
+```
+
+Diagnostics (2) and (3) are analogous except the guard is provided by
+`RuntimeInformation.IsOSPlatformAndEarlier()`.
+
+### Code Fixers
+
+* Suggest wrapping statement in platform guard
+* Suggest annotating the method with an attribute
```

Comment:
What about the scenario where an API is _changed_? This is a case that is going to happen in the future with Xamarin and has happened in the past (with the reflection apis). 

It seems likely to me (but perhaps we won't ever do this?) that we would deprecate an API because there is a safer, better alternative that we want people to move to. Is that scenario out of scope (If so please list it in the non-goals section)

Reply:
Based on my conversation with @mhutch that's not how we envision this moving forward. We don't expect Apple/Android to make ABI breaking changes to the shape of an API (except for removing an API). This means you can always compile against the latest version of an OS API and run on older versions.

Reply:
I would agree with Immo here. I can't remember a time where there was a breaking change. New APIs are added and old ones are obsoleted.

Reply:
Even removals are uncommon, often short-lived mistakes.

Apple _generally_ **rejects** API, e.g. the current `UIWebView` saga, when applications are submitted to the AppStore. Sadly that does not really make it easier for us, but that's a different problem :)